src/org/jetbrains/java/decompiler/modules/decompiler/sforms/SSAProcessor.java

```diff
@@ -0,0 +1,964 @@
+    HashSet<String> updated = new HashSet<>();
+    do {
+      // System.out.println("~~~~~~~~~~~~~ \r\n"+root.toJava());
+      this.ssaStatements(dgraph, updated, false, mt, iteration++);
+      // System.out.println("~~~~~~~~~~~~~ \r\n"+root.toJava());
+    }
+    while (!updated.isEmpty());
+
+    if(this.doLiveVariableAnalysisRound) {
+      this.ssaStatements(dgraph, updated, true, mt, iteration);
+
+      this.ssuversions.initDominators();
+    }
+  }
+
+  private void ssaStatements(DirectGraph dgraph, HashSet<String> updated, boolean calcLiveVars, StructMethod mt, int iteration) {
+
+    DotExporter.toDotFile(dgraph, mt, "ssaStatements_" + iteration, this.outVarVersions);
+
+    for (DirectNode node : dgraph.nodes) {
+
+      updated.remove(node.id);
+      this.mergeInVarMaps(node, dgraph);
+
+      SFormsFastMapDirect varmap = this.inVarVersions.get(node.id);
+      VarMapHolder varmaps = VarMapHolder.ofNormal(varmap);
+
+      if (node.exprents != null) {
+        for (Exprent expr : node.exprents) {
+          varmaps.toNormal(); // make sure we are in normal form
+          this.processExprent(node, expr, varmaps, node.statement, calcLiveVars);
+        }
+      }
+
+      if(this.blockFieldPropagation) {
+        // quick solution: 'dummy' field variables should not cross basic block borders (otherwise problems e.g. with finally loops - usage without assignment in a loop)
+        // For the full solution consider adding a dummy assignment at the entry point of the method
+        boolean allow_field_propagation = node.succs.isEmpty() || (node.succs.size() == 1 && node.succs.get(0).preds.size() == 1);
+
+        if (!allow_field_propagation) {
+          varmaps.getIfTrue().removeAllFields();
+          varmaps.getIfFalse().removeAllFields();
+        }
+      }
+
+      if (this.hasUpdated(node, varmaps)) {
+        this.outVarVersions.put(node.id, varmaps.getIfTrue());
+        if (dgraph.mapNegIfBranch.containsKey(node.id)) {
+          this.outNegVarVersions.put(node.id, varmaps.getIfFalse());
+        }
+
+        // Don't update the node if it wasn't discovered normally, as that can lead to infinite recursion due to bad ordering!
+        if (!dgraph.extraNodes.contains(node)) {
+          for (DirectNode nd : node.succs) {
+            updated.add(nd.id);
+          }
+        }
+      }
+    }
+  }
+
+  private void processExprent(DirectNode node, Exprent expr, VarMapHolder varmaps, Statement stat, boolean calcLiveVars) {
+
+    if (expr == null) {
+      return;
+    }
+
+    assert varmaps.isNormal();
+
+    switch (expr.type) {
+      case Exprent.EXPRENT_IF: {
+        IfExprent ifexpr = (IfExprent) expr;
+        this.processExprent(node, ifexpr.getCondition(), varmaps, stat, calcLiveVars);
+        return;
+      }
+      case Exprent.EXPRENT_ASSIGNMENT: {
+        AssignmentExprent assexpr = (AssignmentExprent) expr;
+        // Kropp : pretty sure ssa(u) gives invalid results if this asserts fails
```

Comment:
We found that this was true- I think you should either fix it here or remove the name tag from the start of the comment